### PR TITLE
DefenderAI that makes octs follow units.

### PR DESCRIPTION
**Pull request by MEEPofFaith**
https://user-images.githubusercontent.com/54301439/108773617-7bfdef80-7513-11eb-810a-9218a85aefcc.mp4

Foreshadow-like targetting
image

### CODE REVIEW

File: core/src/mindustry/ai/types/DefenderAI.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    @Override
+    protected void updateTargeting(){
+        if(retarget()) target = findTarget(unit.x, unit.y, 0f, true, true);
+    }
+    
+    @Override
+    protected Teamc findTarget(float x, float y, float range, boolean air, boolean ground){
+        Sortf sort = (u, tx, ty) -> -u.maxHealth + Mathf.dst2(u.x, u.y, tx, ty) / 800f; //Sort by max health and closer target.
```
Comment (Anuken) [inline core/src/mindustry/ai/types/DefenderAI.java:31]: This should not be defined as a variable - inline it.

File: core/src/mindustry/ai/types/DefenderAI.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import mindustry.entities.units.*;
+import mindustry.gen.*;
+import mindustry.world.meta.*;
+
+public class DefenderAI extends AIController{
+
+    @Override
+    public void updateMovement(){
+        if(target != null){
+            moveTo(target, -1f);
```
Comment (Anuken) [inline core/src/mindustry/ai/types/DefenderAI.java:15]: It should not hug targets. Circle length should be greater than 0 and depend on target hitbox size - octs crashing into eclipses is a bad idea.

File: core/src/mindustry/ai/types/DefenderAI.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+    @Override
+    protected void updateTargeting(){
+        if(retarget()) target = findTarget(unit.x, unit.y, 0f, true, true);
+    }
+    
+    @Override
+    protected Teamc findTarget(float x, float y, float range, boolean air, boolean ground){
+        Sortf sort = (u, tx, ty) -> -u.maxHealth + Mathf.dst2(u.x, u.y, tx, ty) / 800f; //Sort by max health and closer target.
+        return Units.closest(unit.team, x, y, u -> !u.dead() && u.type != unit.type && !u.isPlayer(), sort);
```
Comment (Anuken) [inline core/src/mindustry/ai/types/DefenderAI.java:32]: Why not follow players?

**Review by Anuken — CHANGES_REQUESTED** [review]

File: core/src/mindustry/ai/types/DefenderAI.java (hunk 3 above)
Comment (MEEPofFaith) [reply core/src/mindustry/ai/types/DefenderAI.java:32]: I already changed this back before you sent this message.

**Comment by guncatt** [comment]
I would rather have oct target places where it can cover the most units but this is good

File: core/src/mindustry/ai/types/DefenderAI.java (hunk 2 above)
Comment (Anuken) [reply core/src/mindustry/ai/types/DefenderAI.java]: This still hugs targets, and will crash into Eclipse units.

**Comment by MEEPofFaith** [comment]
> I would rather have oct target places where it can cover the most units but this is good

I'm not gonna bother trying to figure out how to math out how to target the "densest cluster"

**Comment by MEEPofFaith** [comment]
I've resolved the changes why does it still say changes are requested.

**Comment by Anuken** [comment]
I'm still not entirely happy with this implementation, will refine it after merge.

**Pull request merged**